packages/application-shell/src/components/quick-access/quick-access.tsx

```diff
@@ -0,0 +1,508 @@
+          searchProductIdsAction(
+            searchText,
+            applicationContext.project.key,
+            applicationContext.dataLocale
+          )
+        );
+        return result && result.hits ? result.hits.map(hit => hit.id) : [];
+      }
+      return [];
+    },
+    [
+      applicationContext.dataLocale,
+      applicationContext.project,
+      dispatchFetchProductIds,
+    ]
+  );
+
+  const dispatchFetchPimIndexerStatus = useAsyncDispatch<
+    ReturnType<typeof pimIndexerStatusAction>,
+    unknown
+  >();
+  const fetchPimIndexerStatus = React.useCallback<
+    () => Promise<
+      | typeof pimIndexerStates['INDEXED']
+      | typeof pimIndexerStates['NOT_INDEXED']
+    >
+  >(async () => {
+    if (applicationContext.project && applicationContext.dataLocale) {
+      try {
+        dispatchFetchPimIndexerStatus(
+          pimIndexerStatusAction(
+            applicationContext.project.key,
+            applicationContext.dataLocale
+          )
+        );
+        return pimIndexerStates.INDEXED;
+      } catch (error) {
+        // project is not using pim-indexer when response error code is 404,
+        // but we treat all errors as non-indexed as a safe guard, so we're
+        // not checking the response error code at all
+        return pimIndexerStates.NOT_INDEXED;
+      }
+    }
+    return pimIndexerStates.NOT_INDEXED;
+  }, [
+    applicationContext.dataLocale,
+    applicationContext.project,
+    dispatchFetchPimIndexerStatus,
+  ]);
+
+  const getProjectIndexStatus = React.useCallback(async () => {
+    // skip when there is no project
+    if (!applicationContext.project) return pimIndexerStates.NOT_INDEXED;
+
+    const canViewProducts = hasSomePermissions(
+      ['ViewProducts'],
+      applicationContext.permissions
+    );
+
+    // skip checking when user can't view products anyways
+    if (!canViewProducts) return pimIndexerStates.NOT_INDEXED;
+
+    return await fetchPimIndexerStatus();
+  }, [
+    applicationContext.permissions,
+    applicationContext.project,
+    fetchPimIndexerStatus,
+  ]);
+
+  React.useEffect(() => {
+    if (props.pimIndexerState === pimIndexerStates.UNCHECKED) {
+      getProjectIndexStatus().then(status => {
+        if (status) props.onPimIndexerStateChange(status);
+      });
+    }
+
+    return () => {
+      saveHistoryEntries(historyEntries);
+    };
+  });
+
+  const execQuery = React.useCallback<ExecGraphQlQuery>(
+    (Query, variables) =>
+      apolloClient
+        .query({
+          query: Query,
+          errorPolicy: 'ignore',
+          variables,
+        })
+        .then(response => response.data),
+    [apolloClient]
+  );
+
+  const getNextCommands = React.useCallback<
+    (command: Command) => Promise<Command[]>
+  >(
+    async command => {
+      if (!command.subCommands) return [];
+      if (Array.isArray(command.subCommands)) return command.subCommands;
+      return await command.subCommands(execQuery);
```

Comment:
As mentioned in the other comments, this is where the subcommand is called as a function, which needs to return a Promise